<launch>
    <arg name="save_name" default="test"/>
    <group ns="blimp">
        <node name="blimp_controller" pkg="blimp_navigation" type="blimp_controller_node" output="screen">
            <param name="file_name" command="date +'$(find blimp_navigation)/log/%m%d_%H%M-controller_$(arg save_name)_l.csv'"/>
        </node>
        <node name="pid_ux" pkg="pid" type="controller">
            <param name="Kp" value="-40.0" />
            <param name="Ki" value="-0.0" />
            <param name="Kd" value="-300.0" />
            <param name="upper_limit"  value="64.0" />
            <param name="lower_limit"  value="-64.0" />
            <param name="windup_limit" value="64.0" />
            <param name="min_loop_frequency" value="8.0" />
            <param name="max_loop_frequency" value="12.0" />
            <param name="topic_from_controller" value="u_x" />
            <param name="topic_from_plant" value="dist" />
            <param name="reset_pid_topic" value="reset_x" />
            <remap from="setpoint" to="/setpoint_x" />
        </node>
        <node name="pid_uy" pkg="pid" type="controller">
            <param name="Kp" value="-100.0" />
            <param name="Ki" value="-0.0" />
            <param name="Kd" value="-600.0" />
            <param name="upper_limit"  value="128.0" />
            <param name="lower_limit"  value="-128.0" />
            <param name="windup_limit" value="128.0" />
            <param name="min_loop_frequency" value="8.0" />
            <param name="max_loop_frequency" value="12.0" />
            <param name="topic_from_controller" value="u_y" />
            <param name="topic_from_plant" value="dist_y" />
            <remap from="setpoint" to="/setpoint_y" />
            <param name="reset_pid_topic" value="reset_y" />
        </node>
        <node name="pid_uz" pkg="pid" type="controller">
            <param name="Kp" value="-200.0" /> <!-- Tuned, hopefully -->
            <param name="Ki" value="-10.0" />
            <param name="Kd" value="-900.0" />
            <param name="upper_limit"  value="128.0" />
            <param name="lower_limit"  value="-128.0" />
            <param name="windup_limit" value="128.0" />
            <param name="min_loop_frequency" value="8.0" />
            <param name="max_loop_frequency" value="12.0" />
            <param name="topic_from_controller" value="u_z" />
            <param name="topic_from_plant" value="z" />
            <remap from="setpoint" to="/setpoint_z" />
        </node>
        <node name="ble_blimp" pkg="ble_joy_controller" type="ble_pid_node.py" output="screen" > </node>
    </group>
    
    <!--<node name="streamer" pkg="blimp_navigation" type="stream_file_node" />-->
    <!--<node name="streamer" pkg="fisheye_stereo" type="stream_file" />-->
    
    <node name="track_left" pkg="blimp_navigation" type="blimp_human_track_node" output="screen">
        <remap from="camera" to="left"/>
        <param name="detect_human" value="true" />
        <param name="file_name" command="date +'$(find blimp_navigation)/log/%m%d_%H%M-$(arg save_name)_l'"/>
    </node>
    <node name="track_right" pkg="blimp_navigation" type="blimp_human_track_node" output="screen">
        <remap from="camera" to="right"/>
        <param name="detect_human" value="true" />
        <param name="file_name" command="date +'$(find blimp_navigation)/log/%m%d_%H%M-$(arg save_name)_r'"/>
    </node>
    
    <node name="triangulator" pkg="blimp_navigation" type="triangulate_blimp_node">
        <param name="file_name" command="date +'$(find blimp_navigation)/log/%m%d_%H%M-$(arg save_name)'"/>
    </node>
    
    <node pkg="rostopic" type="rostopic" name="set_x" args="pub /setpoint_x std_msgs/Float64 0.0" />
    <node pkg="rostopic" type="rostopic" name="set_y" args="pub /setpoint_y std_msgs/Float64 0.0" />
    <node pkg="rostopic" type="rostopic" name="set_z" args="pub /setpoint_z std_msgs/Float64 0.0" />
    
    <node name="dynamic_reconfigure" pkg="rqt_reconfigure" type="rqt_reconfigure"> </node>
    
    <node name="view_l" pkg="image_view" type="image_view">
        <remap from="image" to="cam_left/detection_image" />
    </node>
    <node name="view_r" pkg="image_view" type="image_view">
        <remap from="image" to="cam_right/detection_image" />
    </node>
</launch>
